### PR TITLE
feat(docs): add support of collapsible sections

### PR DESCRIPTION
**Pull request by k-dovgan**

### CODE REVIEW

**Comment by k-dovgan** [comment]
https://universum.readthedocs.io/en/collapsible-blocks/

File: doc/_ext/collapsible.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from docutils import nodes
+from docutils.parsers.rst import Directive
+from docutils.parsers.rst.directives import unchanged
+
+
+class CollapsibleNode(nodes.General, nodes.Element):
+    pass
+
+
+def visit_collapsible_node(self, node):
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:13]: Function argument named "self" has special meaning in Python.
Please change it.

File: doc/_ext/collapsible.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        collapsible_node = CollapsibleNode('\n'.join(self.content),
+                                           id=randomize_name("collapsible"),
+                                           header=self.options.get("header", "Click here to show/hide"))
+        self.state.nested_parse(self.content, self.content_offset, collapsible_node)
+
+        return [collapsible_node]
+
+
+def setup(app):
+    app.add_node(CollapsibleNode, html=(visit_collapsible_node, depart_collapsible_node))
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:49]: How this looks in non-html output?

File: doc/_ext/collapsible.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    return name + "-" + "".join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6))
+
+
+class CollapsibleDirective(Directive):
+    has_content = True
+    option_spec = {'header': unchanged}
+
+    def run(self):
+        collapsible_node = CollapsibleNode('\n'.join(self.content),
+                                           id=randomize_name("collapsible"),
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:41]: Is it really necessary to use random name instead of some counter?

File: doc/_static/theme_overrides.css [hunk 4]
```diff
@@ -20,5 +20,42 @@
...
+
+    .hide + label + div + .nl {
+        display: block;
+    }
+    .hide:checked + label + div + .nl::after {
+        display: none;
+    }
+
+    .hide + label .sectionLbl::before {
+        content: "[+] ";
```
Comment (i-keliukh) [inline doc/_static/theme_overrides.css:55]: Please change it to look nicer

File: doc/_ext/collapsible.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    <span class="sectionLbl"><b>{node["header"]}</b></span>
+</label>
+<div>
+""")
+
+
+def depart_collapsible_node(self, node):
+    self.body.append("""
+</div>
+<span class="nl">
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:26]: I think you can totally remove nl and sectionLbl if you change the label display mode to "block".

File: doc/_ext/collapsible.py (hunk 1 above)
Comment (k-dovgan) [reply doc/_ext/collapsible.py:12]: These functions are to become class methods; so passing 'self' as a first argument is kind of a common practice as I can see in tutorials.

File: doc/_ext/collapsible.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+class CollapsibleNode(nodes.General, nodes.Element):
+    pass
+
+
+def visit_collapsible_node(self, node):
+    self.body.append(f"""
+<input type="checkbox" id="{node["id"]}" class="hide">
+<label for="{node["id"]}">
+    <span class="collapsible"><b>{node["header"]}</b></span>
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:16]: What is the purpose of this span and its class?

File: doc/_static/theme_overrides.css [hunk 7]
```diff
@@ -20,5 +20,25 @@
...
+
+
+    .hide {
+        display: none;
+    }
+    .hide + label ~ div {
+        display: none;
+    }
+    .hide + label {
+        display: block;
```
Comment (i-keliukh) [inline doc/_static/theme_overrides.css:32]: I propose to add some minor styling here: font size, background color, cursor.
Otherwise it is not clear that this is interactive element.
Also its font is even smaller than that of the normal text.

File: doc/_ext/collapsible.py (hunk 6 above)
Comment (i-keliukh) [reply doc/_ext/collapsible.py:16]: Also, please remove bold element. The visual style should be done in the style sheet, not in html markup.

File: doc/_static/theme_overrides.css [hunk 8]
```diff
@@ -20,5 +20,29 @@
...
+        font-size: 1em;
+        font-weight: bold;
+        padding: 0.5em 1.5em;
+        text-shadow: 2px 2px lightgrey;
```
Comment (i-keliukh) [inline doc/_static/theme_overrides.css:36]: Remove text-shadow. It is a cheap effect from 90s.

File: doc/_static/theme_overrides.css [hunk 9]
```diff
@@ -20,5 +20,29 @@
...
+    }
+    .hide:checked + label + div {
+        display: block;
+    }
+    .hide + label::after {
+        content: " ▶";
+    }
+    .hide:checked + label::after {
+        content: " ▼";
```
Comment (i-keliukh) [inline doc/_static/theme_overrides.css:45]: The "▼" character makes label jump up and down depending on the collapse status.
Please fix this issue.

**Comment by i-keliukh** [comment]
Possible ways of improvement:
1. fix collapse mark jumping after rotation
2. add background to label
3. disable text selection on the label

File: doc/_static/theme_overrides.css [hunk 10]
```diff
@@ -20,5 +20,35 @@
...
+    .hide {
+        display: none;
+    }
+    .hide + label ~ div {
+        display: none;
+    }
+    .hide + label {
+        display: block;
+        font-size: 1em;
+        padding: 0.5em 1.5em;
```
Comment (i-keliukh) [inline doc/_static/theme_overrides.css:34]: Please make the horizontal and vertical padding identical.

File: doc/_ext/collapsible.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    def run(self):
+        global collapsible_id_counter
+        collapsible_node = CollapsibleNode('\n'.join(self.content),
+                                           id="collapsible-" + str(collapsible_id_counter),
+                                           header=self.options.get("header", "Click here to show/hide"))
+        collapsible_id_counter += 1
+        self.state.nested_parse(self.content, self.content_offset, collapsible_node)
+
+        self.state.document.settings.env.config.html_static_path.append(os.path.dirname(__file__))
```
Comment (i-keliukh) [inline doc/_ext/collapsible.py:48]: This line will add the path as many times as there are collapsible directives in the documentation.
Please change it to do only once.

**Comment by i-keliukh** [comment]
This item is not implemented yet:
3. disable text selection on the label

File: doc/conf.py [hunk 12]
```diff
@@ -109,7 +109,8 @@
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ['_static']
 html_context = {
-    'css_files': ['_static/theme_overrides.css']
+    'css_files': ['_static/theme_overrides.css',
+                  '_static/collapsible_overrides.css']
```
Comment (i-keliukh) [inline doc/conf.py:113]: According to https://github.com/sphinx-doc/sphinx/issues/1379 you have to use the following code inside your extension instead of changing project config:
```python
env.app.add_stylesheet(stylesheet)
```

File: doc/_ext/collapsible/collapsible.py [hunk 13]
```diff
@@ -45,7 +45,9 @@ def run(self):
         collapsible_id_counter += 1
         self.state.nested_parse(self.content, self.content_offset, collapsible_node)
 
-        self.state.document.settings.env.config.html_static_path.append(os.path.dirname(__file__))
+        directory = os.path.dirname(__file__)
+        if not directory in self.state.document.settings.env.config.html_static_path:
+            self.state.document.settings.env.config.html_static_path.append(directory)
```
Comment (i-keliukh) [inline doc/_ext/collapsible/collapsible.py:50]: The path must start one level above the extension directory. This is needed to include the extension name when adding css.

File: doc/conf.py (hunk 12 above)
Comment (k-dovgan) [reply doc/conf.py]: I tried using this function in 'app' from `setup (app)` directly, and also in same 'env' we successfully use to add extension contents to '_static', but unlike adding file directly to 'html_context' it didn't work. Needs further investigation.

(Also '.add_stylesheet' was renamed into '.add_css_file')

Comment (i-keliukh) [reply doc/conf.py]: Move adding of theme_overrides.css into setup function in the conf.py, like below:
```
app.add_css_file('theme_overrides.css')
```
And do not set html_context at all.

File: doc/_ext/collapsible/collapsible.py (hunk 13 above)
Comment (i-keliukh) [reply doc/_ext/collapsible/collapsible.py:50]: Unfortunately, the css file has to be placed into separate directory.
With current approach, extension source file is also published into html directory.

File: doc/conf.py (hunk 12 above)
Comment (k-dovgan) [reply doc/conf.py]: Also moved setting 'html_static_path' to `setup()` to allow it to be appended in extensions too.

File: doc/conf.py [hunk 14]
```diff
@@ -101,18 +100,13 @@
...
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
-html_context = {
-    'css_files': ['_static/theme_overrides.css',
-                  '_static/collapsible/collapsible_overrides.css']
-}
-
+# html_static_path = ['_static']
```
Comment (i-keliukh) [inline doc/conf.py:108]: Commented code?

Comment (k-dovgan) [reply doc/conf.py:108]: Well it is above? E.g. in line 103. I thought it all is a part of default config?

File: doc/_ext/collapsible/collapsible.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                                           header=self.options.get("header", "Click here to show/hide"))
+        collapsible_id_counter += 1
+        self.state.nested_parse(self.content, self.content_offset, collapsible_node)
+
+        return [collapsible_node]
+
+
+def add_css_overrides(app):
+    copy_asset(os.path.join(os.path.dirname(__file__), 'css'),
+               os.path.join(app.builder.outdir, '_static'))
```
Comment (i-keliukh) [inline doc/_ext/collapsible/collapsible.py:54]: Could you, please, check that target directory is always "_static"?
My guess is that there is a specific property inside builder object for path to static assets.

Comment (k-dovgan) [reply doc/_ext/collapsible/collapsible.py:54]: I can't see it pointed out directly, but in Sphinx API documentation the '_static' directory is hardcoded to their example and is not in any way configurable:
https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx.application.Sphinx.add_css_file

Comment (i-keliukh) [reply doc/_ext/collapsible/collapsible.py:54]: Looking into sources of sphinx html builder, I can confirm that "_static" is hardcoded there multiple times.

File: doc/index.rst [hunk 16]
```diff
@@ -33,4 +33,19 @@ Sometimes `Universum` system can be referred to as the framework or just CI.
 See more about :doc:`installing Universum <install>`, :doc:`creating a configuration file <configuring>` and
 :doc:`parametrizing Universum execution <args>`, or read the :doc:`step-by-step instruction on getting started <guide>`.
 
+
```
Comment (i-keliukh) [inline doc/index.rst:36]: Please remove this

File: setup.py [hunk 17]
```diff
@@ -16,7 +16,7 @@ def readme():
 
 vcs = p4 + github
 
-docs = ('sphinx', 'sphinx-argparse', 'sphinx_rtd_theme')  # This extra is required for RTD to generate documentation
+docs = ('sphinx==3.4.3', 'sphinx-argparse', 'sphinx_rtd_theme')  # This extra is required for RTD to generate documentation
```
Comment (i-keliukh) [inline setup.py:19]: unrelated change

File: tests/test_p4_exception_handling.py [hunk 18]
```diff
@@ -3,7 +3,7 @@
 import pytest
 
 from universum import __main__
-from tests.perforce_utils import P4Environment
+from .perforce_utils import P4Environment
```
Comment (i-keliukh) [inline tests/test_p4_exception_handling.py:6]: unrelated change

File: tests/utils.py [hunk 19]
```diff
@@ -46,7 +46,7 @@ def is_pycharm():
 
 
 def randomize_name(name):
-    return name + "-".join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6))
+    return name + "-" + "".join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6))
```
Comment (i-keliukh) [inline tests/utils.py:49]: unrelated change

**Review by i-keliukh — APPROVED** [review]
